File: algorithms/47. Permutations II/solution.cc
class Solution {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<vector<int>> result;
        vector<int> snums(nums);
        sort(snums.begin(), snums.end());
        dfs(0, snums, result);
        return result;
    }

    void dfs(int start, vector<int> nums, vector<vector<int>> & result) {
        if ( start == nums.size() ) {
            result.push_back(nums);
        }
        for ( int i = start; i < nums.size(); ++i ) {
            if ( i == start || nums[i] != nums[start] ) {
                swap(nums[i], nums[start]);
                dfs(start + 1, nums, result);
            }
        }
    }
};

File: algorithms/47. Permutations II/solution.py
class Solution(object):
    def permuteUnique(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = [[]]
        for n in nums:
            newres = []
            for p in res:
                for i in xrange(len(p) + 1):
                    newres.append(p[:i] + [n] + p[i:])
                    if i < len(p) and p[i] == n:
                        break
            res = newres
        return res

# 30 / 30 test cases passed.
# Status: Accepted
# Runtime: 119 ms
# 58.90 %
